### PR TITLE
[PoC] perf(ASGI): swap out App.__call__ for a potentially more optimized one after app startup

### PR DESCRIPTION
**Pull request by vytas7**
This is a prototype. Things to do:

- [ ] Decide whether this is a good idea
- [ ] Reason about `App` subclassing patterns and issues we might run into
- [ ] Restore 100% coverage
- [ ] Add `NOTE`s (and potentially docs) describing the effect of this hack

### CODE REVIEW

**Comment by kgriffs** [comment]
Can we quantify the performance gain from doing this?

File: falcon/asgi/app.py [hunk 1]
```diff
@@ -270,18 +271,44 @@ async def process_response(self, req, resp, resource, req_succeeded)
...
+            if app:
+                if app._lifespan_status != EventType.LIFESPAN_STARTUP_COMPLETE:
+                    all_started = False
+            else:
+                disposed.add(ref)
+
+        cls._known_instances.difference_update(disposed)
+
+        if all_started:
+            cls.__call__ = cls._call_impl
```
Comment (kgriffs) [inline falcon/asgi/app.py:297]: > Reason about App subclassing patterns and issues we might run into

Indeed, we will need to add test cases to cover these patterns.

**Comment by vytas7** [comment]
Re performance gain, we are shaving off one compatibility-`await` (and the real gain is for some strange reason slightly even more :thinking: ).

**Comment by vytas7** [comment]
There is unfortunately some puzzling, unsolved issue with cythonized stuff and "real" ASGI servers that I don't understand.

Directly involving the cythonized app seems to produce a noticeable speedup, but wrapping that in a pure Python coroutine function negates that advantange, and even turns it into a slowdown. It seems that at least `uvicorn` always invokes the app in the way that it is unfortunately slowed down.

**Comment by vytas7** [comment]
I'll kill this for now until the above is demystified.

I was testing Cython 3.0 alpha, and while it is also disappointing, it performed slightly better, so it might well be an inherent limitation of cythonized generators for now :disappointed:

**Pull request closed without merging**